### ISSUE TITLE
Need guidance on how to customize config file

### ISSUE THREAD

**Issue by patcoll** [docker/distribution#385]
Explain how to:

- Change the location where registry looks for config file (via passing in argument with `docker run`)
- Mount a volume at that location where the config file can be found

This stuff seems obvious for those who have used Docker before, but would be useful for beginners/novices.

**Comment by stevvooe** [comment]
@patcoll Thank you for the report!

@moxiegirl, PTAL but we may want to tweak how the image works to make this a little easier.

**Comment by moxiegirl** [comment]
@patcoll Thanks for the report.  I could see this putting this as an example in the documentation.  There are several candidates for locations in the docs for this. To help me decide, on which page were you looking for this? For example, do you think it belongs here: http://docs.docker.com/registry/deploying/ or on the configuring page?

**Comment by patcoll** [comment]
@moxiegirl Thanks for following up. I was reading both "Deploying" and "Configure" pages at the same time, trying to find where this information was. I opened the "Deploying" page first. I suggest putting an example of using a custom configuration file there (or at least linking to it from there).

It occurs to me that labeling the link "Configure" may be a misleading term, because it suggests that it provides instructions for the action of configuring the registry. As of this writing, the page serves as more of a configuration reference, so I suggest labeling every link to that page as "registry 2.0 configuration reference" or similar.

Perhaps a cookbook- or guide-style set of instructions would help. Section I could imagine existing:

- "Differences from registry 1.0"
- "Mounting data volume for persisting image storage"
- "Customizing a registry 2.0 configuration file"

I'd be willing to contribute if we agree that style would be needed.

**Comment by moxiegirl** [comment]
Overall, I think these suggestions are strong. I'd love it if you would like to contribute. The engineers also review this so I'm not the only reviewer.   Before you begin, the only thoughts I had were these:

_Differences from registry 1.0_  sounds to me like it would fit in the overview as a general piece. But I could be misunderstanding -- do you mean more like "configuration differences between 1.0 and 2.0" as in a specific list of options that have changed? Either way I think it could work --- or maybe both. What do you think?

Good suggestion, let's include `Configure a registry > Configuration reference` (actually I dope slapped myself).

Finally, just fair warning, the bulleted list of headings you propose work well as interior headings on the page. If we want any of them to show up on the menu, we might have to tweak them for length. 

We really appreciate the thoughtful issue and the detailed suggestion.  Let us know if you want to proceed on a PR.

**Comment by KingScooty** [comment]
:+1: 

I've been using docker on a few projects now, and just started looking into setting up a private registry using Registry 2.0. I must admit, i'm finding the existing docs a little bit confusing. I'd really like to see a guide doc that describes in a little more detail how to configure for production and deploy in more of a step by step format, instead of making rather bold assumptions on base level knowledge/familiarity with the registry 2.0 project.

For instance, it actually took me longer than i'd like to admit just finding where the config file is stored. I eventually found where it was by reading the TLS section of the deploy doc.

**Comment by moxiegirl** [comment]
@KingScooty Thank you for the feedback.  This is our first registry 2.0 release and  iteration on the docs.  Given our time/resource constraints, we actually went with a minimal documentation approach. So your feedback is very useful.

Creating a precise step-wise procedure for "deploying to production" is difficult because each company's environment is different.  Writing such a single procedure would require solving a [ "moderately-structured" or "ill-structured" problem](http://en.wikiversity.org/wiki/Introduction_to_Ill-Structured_Problems) for a lot of different use cases.  I can however do several things to help people solve "deploy to production" for their particular use case.

One thing that is useful, is to have several small step-by-step examples of different types of basic deployment configurations.   We do have a start on that here (http://docs.docker.com/registry/deploying/) &mdash. if you don't see your use case there, feel free to suggest another one.  The other thing we can do is find the common elements among all these examples and, as you suggest, create a guidance document that explains the principles and decisions that someone would need to make around deploying to production.

**Comment by jp-gouigoux** [comment]
I agree that having procedure for the complete deployment is difficult because it depends much on the context, but what is missing may be only some punctual information. For example, there is a sentence in the documentation stating "You do this by specifying options on the command line or, more typically, by writing a registry configuration file. The configuration file is in YAML format."... and then nothing ! The text stops... I was expecting to know where was this file, how to pass it, etc.

Same thing at the top of https://docs.docker.com/registry/configuration/ : it states "You configure a registry server using a YAML file.", but nothing in the page about the way to instruct the registry to use another one...

I tried passing a file with absolute or relative format to the command line, but got an error. I looked for a variable that would let me pass the filename but could not find it in the documentation either. I looked at the Dockerfile wondering if the CMD value would help me to figure out how to do it, but nope. I wondered if passing variables for each of the configuration unit would be a possibility, but it only allows to overwrite existing values, and not get rid of, say, filesystem storage to configure s3 storage.

If you can answer on this issue about the right way to pass a custom config file, I would be more than happy to test the different methods, validate them and send a pull request for the "Registry Configuration Reference" documentation page.

**Comment by rsslldnphy** [comment]
I'm also trying to work out how to do this. The [docs](http://docs.docker.com/registry/deploying/) talk about cloning the distribution repository and directly editing the example config file, but I'd like to just use the official image and pass config to it rather than maintaining my own fork... I'm sure it's something simple but I don't know enough yet to work it out.

In short I want to pass a custom config.yml file to the official registry image, rather than building my own. Anyone able to point me in the right direction?

**Comment by rsslldnphy** [comment]
Scratch that, I think I've worked it out:

Add your config file to a directory on the host machine, say as `/opt/registry/config.yml`.
When running the registry, set up this directory as a volume, then pass the filename as the command to be run. So you run command becomes something like:

```
docker run -p 5000:5000 -v /opt/registry/config.yml:/opt/registry/config.yml registry:2.0 /opt/registry/config.yml
```

I'm sure there's better/cleaner ways to do it but that seems to be working for me, in case it helps anyone else.

**Comment by moxiegirl** [comment]
I've been pulled onto another project but I think I should try to take a couple of hours to implement these suggestions. @stevvooe Most likely I'll have to do this on this weekend. Ping me if you want to tweak or add to any of these suggestions.  Thanks @rsslldnphy @jp-gouigoux  for adding to the original commentary.

**Comment by Yantrio** [comment]
I ended up setting environment variables to override the specific settings I wanted
for example 
```
docker run -e "REGISTRY_STORAGE=filesystem"
                 -e "REGISTRY_STORAGE_FILESYSTEM_ROOTDIRECTORY=/tmp/registry" 
                 -p 5000:5000 -p 5001:5001 
                 -v /opt/registry:/tmp/registry 
                 registry:2.0
```

**Comment by dmp42** [comment]
The documentation have been revamped through #553 and #615 

That should now explicitly cover:

 * data location through volume mounts
 * overriding configuration through environment
 * and de-emphasize the need to mount your own config file

If you think this helps addressing the pain points listed in this thread, then we could close this.

cc @stevvooe

**Comment by stefanfoulis** [comment]
I spent a lot of time confused, because not all my settings set by environment variables where being picked up. Some worked (redis stuff) and some where plainly ignored (auth.htpasswd).
Only after digging in the source did I find out, that it is only possible to override settings by environment var that are also set in the config file. And the default config file at ``cmd/registry/config.yml`` only contains a small subset.

IMHO either this should clearly be documented, or the default config file should allow all settings to be set as environment variables.

**Comment by drocsid** [comment]
Do we really need to mount a volume to set the config? It would be nice if we could do this quickly through the command line / docker run, in an uncomplicated manor. In my example I was trying to override my custom build config. I can try some of the examples above, but it would be nice if we could easily pass the config without needing so many steps E.G. setting environment variables & mounting another volume.

I would think this would make (human) testing the distribution configuration much easier.

Here's what I tried:

I was trying to set the config file using docker run [CMD]:

sudo docker run -i -t regis3.0 /home/colin/code/distribution/cmd/registry/cong-dev.yml
configuration error: open /home/colin/code/distribution/cmd/registry/config-dev.yml: no such file or directory
usage: registry <config>
  -test.bench="": regular expression to select benchmarks to run
  -test.benchmem=false: print memory allocations for benchmarks
  -test.benchtime=1s: approximate run time for each benchmark
  -test.blockprofile="": write a goroutine blocking profile to the named file after execution
  -test.blockprofilerate=1: if >= 0, calls runtime.SetBlockProfileRate()
  -test.coverprofile="": write a coverage profile to the named file after execution
  -test.cpu="": comma-separated list of number of CPUs to use for each test
  -test.cpuprofile="": write a cpu profile to the named file during execution
  -test.memprofile="": write a memory profile to the named file after execution
  -test.memprofilerate=0: if >=0, sets runtime.MemProfileRate
  -test.outputdir="": directory in which to write profiles
  -test.parallel=1: maximum test parallelism
  -test.run="": regular expression to select tests and examples to run
  -test.short=false: run smaller test suite to save time
  -test.timeout=0: if positive, sets an aggregate time limit for all tests
  -test.v=false: verbose: print additional output
  -version=false: show the version and exit

Then I realized it was looking for the file in the volume. I thought maybe I could just pass the configuration easily like so:

sudo docker run -i -t regis3.0 $(cat /home/colin/code/distribution/cmd/registry/config-dev.yml)
configuration error: open version:: no such file or directory
usage: registry <config>
  -test.bench="": regular expression to select benchmarks to run
  -test.benchmem=false: print memory allocations for benchmarks
  -test.benchtime=1s: approximate run time for each benchmark
  -test.blockprofile="": write a goroutine blocking profile to the named file after execution
  -test.blockprofilerate=1: if >= 0, calls runtime.SetBlockProfileRate()
  -test.coverprofile="": write a coverage profile to the named file after execution
  -test.cpu="": comma-separated list of number of CPUs to use for each test
  -test.cpuprofile="": write a cpu profile to the named file during execution
  -test.memprofile="": write a memory profile to the named file after execution
  -test.memprofilerate=0: if >=0, sets runtime.MemProfileRate
  -test.outputdir="": directory in which to write profiles
  -test.parallel=1: maximum test parallelism
  -test.run="": regular expression to select tests and examples to run
  -test.short=false: run smaller test suite to save time
  -test.timeout=0: if positive, sets an aggregate time limit for all tests
  -test.v=false: verbose: print additional output
  -version=false: show the version and exit
